src/CommunityToolkit.Maui/Views/Popup/PopupService.cs

```diff
@@ -0,0 +1,163 @@
+	/// </summary>
+	public PopupService()
+	{
+		serviceProvider = IPlatformApplication.Current?.Services
+		                  ?? throw new InvalidOperationException("Could not locate IServiceProvider");
+
+		dispatcher = Microsoft.Maui.Controls.Application.Current?.Dispatcher
+		             ?? throw new InvalidOperationException("Could not locate IDispatcher");
+	}
+
+	internal static void AddPopup<
+		[DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicConstructors)] TPopupView>(
+		IServiceCollection services, ServiceLifetime lifetime)
+		where TPopupView : IView
+	{
+		viewModelToViewMappings.TryAdd(typeof(TPopupView), typeof(TPopupView));
+		Routing.RegisterRoute(typeof(TPopupView).FullName, typeof(TPopupView));
+
+		services.Add(new ServiceDescriptor(typeof(TPopupView), typeof(TPopupView), lifetime));
+	}
+
+	internal static void AddPopup<
+		[DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicConstructors)] TPopupView,
+		[DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicConstructors)] TPopupViewModel>(
+		IServiceCollection services, ServiceLifetime lifetime)
+		where TPopupView : IView
+		where TPopupViewModel : notnull
+	{
+		viewModelToViewMappings.TryAdd(typeof(TPopupViewModel), typeof(TPopupView));
+		Routing.RegisterRoute(typeof(TPopupViewModel).FullName, typeof(TPopupView));
+
+		services.TryAdd(new ServiceDescriptor(typeof(TPopupView), typeof(TPopupView), lifetime));
+		services.TryAdd(new ServiceDescriptor(typeof(TPopupViewModel), typeof(TPopupViewModel), lifetime));
+	}
+
+	internal static void AddPopup<
+		[DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicConstructors)] TPopupView,
+		[DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicConstructors)] TPopupViewModel>(
+		TPopupView popup, TPopupViewModel viewModel, IServiceCollection services, ServiceLifetime lifetime)
+		where TPopupView : IView
+		where TPopupViewModel : notnull
+	{
+		viewModelToViewMappings.TryAdd(typeof(TPopupViewModel), typeof(TPopupView));
+		Routing.RegisterRoute(typeof(TPopupViewModel).FullName, typeof(TPopupView));
+
+		services.TryAdd(new ServiceDescriptor(typeof(TPopupView), (_) => popup, lifetime));
+		services.TryAdd(new ServiceDescriptor(typeof(TPopupViewModel), (_) => viewModel, lifetime));
+	}
+
+	/// <summary>
+	/// 
+	/// </summary>
+	/// <typeparam name="TBindingContext"></typeparam>
+	/// <param name="options"></param>
+	/// <param name="cancellationToken"></param>
+	/// <returns></returns>
+	public async Task<PopupResult> ShowPopupAsync<TBindingContext>(PopupOptions options, CancellationToken cancellationToken)
+		where TBindingContext : notnull
+	{
+		var bindingContext = serviceProvider.GetRequiredService<TBindingContext>();
+		var popupContent = GetPopupContent(bindingContext);
+
+		var navigation = Application.Current?.Windows[^1].Page?.Navigation ?? throw new InvalidOperationException("Unable to get navigation");
```

Comment:
What's your intention with this line of code?
Do you know this isn't the way to get the focused Window?

Reply:
it's a good question. I haven't found a way to get the focused window, assuming the latest window is main. Do you know the API to get the current window?

Reply:
So, all controls have a `Window` property you may take the Window from the parent control, and with that Window you can get the Navigation

Reply:
We do not have a control or parent as it is called from the PopupService